src/v/kafka/server/fetch_memory_units.cc

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/fetch_memory_units.h"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/util/later.hh>
+
+namespace kafka {
+
+fetch_memory_units_manager::fetch_memory_units_manager(
+  ssx::semaphore& kafka_units,
+  ssx::semaphore& fetch_units,
+  local_instance_fn&& local_fn)
+  : _kafka_units(kafka_units)
+  , _fetch_units(fetch_units)
+  , _release_units_timer([this] { release_units(); })
+  , _local_instance_fn(std::move(local_fn)) {
+    _release_units_timer.arm(max_release_period);
+}
+
+ss::future<> fetch_memory_units_manager::stop() {
+    _release_units_timer.cancel();
+    _units_to_release.clear();
+    return ss::now();
+}
+
+void fetch_memory_units_manager::units::adopt(
+  fetch_memory_units_manager::units&& o) {
+    // Adopts assert internally that the units are from the same semaphore.
+    // So there is no need to assert that they are from the same shard here.
+    kafka_units.adopt(std::move(o.kafka_units));
+    fetch_units.adopt(std::move(o.fetch_units));
+}
+
+void fetch_memory_units_manager::release_units() {
+    _release_units_timer.cancel();
+    _release_units_timer.arm(max_release_period);
+
+    _units_to_release.clear();
+}
+
+fetch_memory_units_manager::units&
+fetch_memory_units_manager::units::operator=(units&& o) noexcept {
+    if (this != &o) {
+        this->~units();
+        new (this) units(std::move(o));
+    }
+    return *this;
+}
+
+fetch_memory_units_manager::units::~units() noexcept {
+    if (shard == ss::this_shard_id() || !has_units()) {
+        return;
+    }
+
+    auto f = ss::smp::submit_to(
+      shard,
+      [uk = std::move(kafka_units),
+       uf = std::move(fetch_units)]() mutable noexcept {
+          uk.return_all();
+          uf.return_all();
+      });
+
+    if (!f.available()) {
+        ss::engine().run_in_background(std::move(f));
```

Comment:
Why this instead of just ssx::background above? I don't think it provide any benefit here: we don't need to wait for these futures to resolve before stopping the reactor.